pallets/gear-builtin/src/staking.rs

```diff
@@ -0,0 +1,198 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! # Gear BuiltIn Actors Pallet
+//!
+//! The BuiltIn Actors pallet provides a set of unique accounts that are treated as built-in
+//! actors ids (`ProgramId`).
+//!
+//! - [`Config`]
+//!
+//! ## Overview
+//!
+//! The pallet implements a builtin actor that handles messages from the message queue that
+//! request some sort of staking-related logic (like tokens bonding, validators nominating etc.).
+//!//!
+//! The pallet defines a type `Actor` that implements the [crate::BuiltinActor] trait so that it
+//! can be plugged into the `Gear` pallet to intercept messages popped from the queue.
+//!
+//! The staking built-in actor does three things:
+//! 1. Decodes the message payload into a respective known type (the types are deinfed in the
+//! [gbuiltin-staking] crate);
+//! 2. Based on the action encoded in the message, created a respective dispatchable call from
+//! the `Staking` pallet and checks whether the gas limit is sufficient to execute the call;
+//! 3. Dispatches the call and prepares the response payload, if any.
+//!
+//! In case of an error (decoding, gas insufficiency or dispatch error) the actor returns the
+//! error type that is common for all built-in actors: the [crate::BuiltinActorError].
+
+use super::*;
+use common::Origin;
+use core::marker::PhantomData;
+use frame_support::dispatch::{extract_actual_weight, GetDispatchInfo};
+use gbuiltin_staking::*;
+use pallet_staking::RewardDestination;
+use parity_scale_codec::{Decode, Encode};
+use sp_runtime::traits::{Dispatchable, StaticLookup, UniqueSaturatedInto};
+
+pub struct Actor<T: Config>(PhantomData<T>);
+
+impl<T: Config> Actor<T>
+where
+    T::AccountId: Origin,
+{
+    fn dispatch_call(
+        origin: T::AccountId,
+        call: <T as Config>::RuntimeCall,
+        gas_limit: u64,
+    ) -> (Result<(), BuiltinActorError>, u64) {
+        let call_info = call.get_dispatch_info();
+
+        // Necessary upfront gas sufficiency check
+        if gas_limit < call_info.weight.ref_time() {
+            return (Err(BuiltinActorError::InsufficientGas), 0_u64);
+        }
+
+        // Execute call
+        let res = call.dispatch(frame_system::RawOrigin::Signed(origin).into());
+        let actual_gas = extract_actual_weight(&res, &call_info).ref_time();
+        match res {
+            Ok(_post_info) => {
+                log::debug!(
+                    target: LOG_TARGET,
+                    "Call dispatched successfully",
+                );
+                (Ok(()), actual_gas)
+            }
+            Err(e) => {
+                log::error!(target: LOG_TARGET, "Error disptaching call: {:?}", e);
+                (
+                    Err(BuiltinActorError::Custom(
+                        DispatchErrorReason::RuntimeError.to_string().into(),
```

Comment:
that usage seems supporting my initial idea that this type is unnecessary

Reply:
What's wrong with having a dedicated type as a wrapper around some string literal? We have similar error types all over the place, don't we?